### PR TITLE
Configure team nominate label in prioritization alerts on Zulip

### PR DESCRIPTION
**Pull request by apiraino**
This patch tries to customize the prioritization Zulip notification (`I-{team}-nominated`) according to the team labels on the issue.

It's not complete, there are some cases not covered but I didn't feel complicating too much the logic.

And this is a brainstoming for @fmease that suggested doing so. Feel free to own this patch and bring it to completion (unsure you have permissions on this repo) or suggest how you would like the implementation to be. 

This patch completes https://github.com/rust-lang/rust/pull/119959#pullrequestreview-1820272630

Thanks!

### CODE REVIEW

**Comment by apiraino** [comment]
hey @fmease I was looking into completing this small change.

wdyt? Does it look like what you wanted to accomplish?

**Comment by fmease** [comment]
It definitely does what I had envisioned, thanks! 😊 The major issue I see is that this PR adds very rust-lang/rust-specific logic to triagebot. While triagebot still contains a lot of things specific to rust-lang/rust, I think the general vision for this bot is to move anything repo-specific to configs & to make it repo-agnostic.

cc @WaffleLapkin

**Comment by fmease** [comment]
And I really don't know how one would best abstract over this feature and make it configurable.
If the maintainers of this repo are fine with merging this as is, then that's awesome, if they aren't we need to change the approach.

@Mark-Simulacrum, what do you think?

File: src/handlers/notify_zulip.rs [hunk 1]
```diff
@@ -172,3 +173,69 @@ pub(super) async fn handle_input<'a>(
 
     Ok(())
 }
+
+fn get_team_to_be_nominated(labels: &Vec<Label>, msg: String) -> String {
```
Comment (Mark-Simulacrum) [inline src/handlers/notify_zulip.rs:177]: This doesn't sound like "get", more of a "replace" or "substitute"?

File: src/handlers/notify_zulip.rs [hunk 2]
```diff
@@ -172,3 +173,69 @@ pub(super) async fn handle_input<'a>(
...
+    // - If multiple team labels are found and one of them is "compiler", pick that one
+    //   (currently the only team handling these Zulip notification)
+    // - else, return a generic text
+    if teams.len() == 1 {
+        msg.replace("{team}", teams[0])
+    } else if teams.contains(&"compiler") {
+        msg.replace("{team}", "compiler")
+    } else {
+        msg.replace(
+            "Needs `I-{team}-nominated`?",
```
Comment (Mark-Simulacrum) [inline src/handlers/notify_zulip.rs:194]: This is hardcoding a non-replacement string -- should we really do that? I would expect this trailing case to be just `msg` I think, i.e., no replacement is performed.

**Comment by WaffleLapkin** [comment]
@fmease all zulip stuff is highly specific to rust-lang. so while I'd like it to be more configurable, this specific change does not really change much.

File: src/handlers/notify_zulip.rs (hunk 2 above)
Comment (apiraino) [reply src/handlers/notify_zulip.rs:194]: > This is hardcoding a non-replacement string -- should we really do that? I would expect this trailing case to be just `msg` I think, i.e., no replacement is performed.

I thought a reasonable default message would be better than ``Needs `I-{team}-nominated`?`` which feels like someone forgot to localize a string :-)

But I am also ok with leaving it as it is. I'm sure nobody reads it anyway :smile:

**Comment by apiraino** [comment]
> The major issue I see is that this PR adds very rust-lang/rust-specific logic to triagebot.

That's true, I agree. However making it _really_ generic is beyond the time I want to allocate on implementing this change (see also my first comment). So if it's not _that_ bad and it's accepted, great, otherwise I'll have to drop the mic on this :smile:

**Review by Mark-Simulacrum — APPROVED** [review]

File: src/handlers/notify_zulip.rs [hunk 3]
```diff
@@ -172,3 +173,66 @@ pub(super) async fn handle_input<'a>(
 
     Ok(())
 }
+
+fn replace_team_to_be_nominated(labels: &Vec<Label>, msg: String) -> String {
```
Comment (fmease) [inline src/handlers/notify_zulip.rs:177]:
```suggestion
fn replace_team_to_be_nominated(labels: &[Label], msg: String) -> String {
```

File: src/handlers/notify_zulip.rs [hunk 4]
```diff
@@ -172,3 +173,66 @@ pub(super) async fn handle_input<'a>(
...
+        .map(|label| &label.name)
+        .filter_map(|label| label.strip_prefix("T-"))
+        .collect::<Vec<&str>>();
+
+    // - If a single team label is found, replace the placeholder with that one
+    // - If multiple team labels are found and one of them is "compiler", pick that one
+    //   (currently the only team handling these Zulip notification)
+    // - else, do nothing
+    if teams.len() == 1 {
+        msg.replace("{team}", teams[0])
```
Comment (fmease) [inline src/handlers/notify_zulip.rs:189]:
```suggestion
    if let [team] = &*teams {
        msg.replace("{team}", team)
```

File: src/handlers/notify_zulip.rs [hunk 5]
```diff
@@ -172,3 +173,66 @@ pub(super) async fn handle_input<'a>(
...
+    } else if teams.contains(&"compiler") {
+        msg.replace("{team}", "compiler")
+    } else {
+        msg
+    }
+}
+
+#[test]
+fn test_notification() {
+    let mut msg = replace_team_to_be_nominated(&vec![], "Needs `I-{team}-nominated`?".to_string());
```
Comment (fmease) [inline src/handlers/notify_zulip.rs:199]:
```suggestion
    let mut msg = replace_team_to_be_nominated(&[], "Needs `I-{team}-nominated`?".to_string());
```

ditto et seqq.

**Review by fmease — COMMENTED** [review]
small stylistic nits, not blocking. sorry, forgot to submit a code review earlier.

File: src/handlers/notify_zulip.rs (hunk 4 above)
Comment (apiraino) [reply src/handlers/notify_zulip.rs]: nit: I find the suggested syntax a bit obscure. What is the difference between the two versions? IIUC avoiding a `.len()` saves a little memory? Like one an allocation less? (trying to learn something :slightly_smiling_face: )

**Comment by apiraino** [comment]
applied the suggested nits!

thanks @fmease for your time :)

File: src/handlers/notify_zulip.rs (hunk 4 above)
Comment (fmease) [reply src/handlers/notify_zulip.rs]: I see what you mean (esp. the `&*` which I sometimes replace with `.as_slice()` for readability).

It's not about allocations but panicking. By using a slice pattern, one can quickly see that this code cannot panic (unless `-Zoom=panic` ofc ^^) but with the check and the explicit indexing, it's “less obvious”. In this case, admittedly it doesn't make a huge difference but it's a nice pattern to follow. It allows for [local reasoning](https://stackoverflow.com/questions/61207079/what-is-local-reasoning-exactly). You wouldn't write `if x.is_some() { x.unwrap() }` either for example.

**Review by fmease — APPROVED** [review]

File: src/handlers/notify_zulip.rs (hunk 4 above)
Comment (apiraino) [reply src/handlers/notify_zulip.rs]: awesome, TIL! Thanks for the additional context

**Comment by fmease** [comment]
needs rebase

**Review by ehuss — APPROVED** [review]
Thanks! I'm going to go ahead and merge based on previous reviews, and since this seems pretty simple.

@apiraino Do you mind following up and post an update to the documentation at https://github.com/rust-lang/rust-forge/blob/master/src/triagebot/zulip-notifications.md to add a brief mention of this?

**Pull request merged**